From 9e1cd8c17905bda08b39ac5c3349f3b0eb962210 Mon Sep 17 00:00:00 2001
From: vulsmike <mikelis.vuls@intel.com>
Date: Fri, 23 Feb 2018 12:57:01 +0200
Subject: [PATCH 3/3] add_intel_specific_to_nl80211

---
 include/uapi/linux/nl80211.h | 32 ++++++++++++++++++++++++++++++++
 1 file changed, 32 insertions(+)

diff --git a/include/uapi/linux/nl80211.h b/include/uapi/linux/nl80211.h
index d3cbe48..91a6301 100644
--- a/include/uapi/linux/nl80211.h
+++ b/include/uapi/linux/nl80211.h
@@ -1936,6 +1936,10 @@ enum nl80211_commands {
  *	attribute.
  * @NL80211_ATTR_NAN_MATCH: used to report a match. This is a nested attribute.
  *	See &enum nl80211_nan_match_attributes.
+ * @NL80211_ATTR_VENDOR_WDS: used to enable the wds vendor specific capability
+ * @NL80211_ATTR_STA_RSSI: Received signal strength indication (u32)
+ * @NL80211_ATTR_RADAR_BIT_MAP: Bit map of channel where radar was detected
+ * @NL80211_ATTR_SB_DFS_BW: Sub Band DFS new operation bandwidth
  *
  * @NL80211_ATTR_BSSID: The BSSID of the AP. Note that %NL80211_ATTR_MAC is also
  *	used in various commands/events for specifying the BSSID.
@@ -2341,6 +2345,14 @@ enum nl80211_attrs {
 
 	NL80211_ATTR_BSSID,
 
+	NL80211_ATTR_VENDOR_WDS,
+	NL80211_ATTR_STA_RSSI,
+	NL80211_ATTR_RADAR_BIT_MAP,
+	NL80211_ATTR_SB_DFS_BW,
+	NL80211_ATTR_HE_CAPABILITY,
+	NL80211_ATTR_HE_CAPABILITY_MASK,
+	NL80211_ATTR_HE_BSS_COLOR,
+
 	/* add attributes here, update the policy in nl80211.c */
 
 	__NL80211_ATTR_AFTER_LAST,
@@ -2387,6 +2399,11 @@ enum nl80211_attrs {
 #define NL80211_TKIP_DATA_OFFSET_RX_MIC_KEY	24
 #define NL80211_HT_CAPABILITY_LEN		26
 #define NL80211_VHT_CAPABILITY_LEN		12
+#define NL80211_HE_CAPABILITY_LEN		15
+#define NL80211_HE_CAPABILITY_OFFSET		0
+#define NL80211_HE_CAPABILITY_MAX_LEN		64
+#define NL80211_HE_CAPABILITY_MCS_NSS_OFFSET	15
+#define NL80211_HE_CAPABILITY_MCS_NSS_LEN	8
 
 #define NL80211_MAX_NR_CIPHER_SUITES		5
 #define NL80211_MAX_NR_AKM_SUITES		2
@@ -2698,6 +2715,7 @@ enum nl80211_sta_info {
 	NL80211_STA_INFO_TID_STATS,
 	NL80211_STA_INFO_RX_DURATION,
 	NL80211_STA_INFO_PAD,
+	NL80211_STA_INFO_MAX_RSSI,
 
 	/* keep last */
 	__NL80211_STA_INFO_AFTER_LAST,
@@ -2797,6 +2815,9 @@ enum nl80211_mpath_info {
  * @NL80211_BAND_ATTR_VHT_MCS_SET: 32-byte attribute containing the MCS set as
  *	defined in 802.11ac
  * @NL80211_BAND_ATTR_VHT_CAPA: VHT capabilities, as in the HT information IE
+ * @NL80211_BAND_ATTR_HE_MCS_SET: 32-byte attribute containing the MCS set as
+ *	defined in 802.11ax
+ * @NL80211_BAND_ATTR_HE_CAPA: HE capabilities, as in the HT information IE
  * @NL80211_BAND_ATTR_MAX: highest band attribute currently defined
  * @__NL80211_BAND_ATTR_AFTER_LAST: internal use
  */
@@ -2812,6 +2833,9 @@ enum nl80211_band_attr {
 
 	NL80211_BAND_ATTR_VHT_MCS_SET,
 	NL80211_BAND_ATTR_VHT_CAPA,
+  
+	NL80211_BAND_ATTR_HE_MCS_SET,
+	NL80211_BAND_ATTR_HE_CAPA,
 
 	/* keep last */
 	__NL80211_BAND_ATTR_AFTER_LAST,
@@ -5128,4 +5152,12 @@ enum nl80211_nan_match_attributes {
 	NL80211_NAN_MATCH_ATTR_MAX = NUM_NL80211_NAN_MATCH_ATTR - 1
 };
 
+enum nl80211_sb_dfs_bw {
+	NL80211_SB_DFS_BW_NORMAL,
+	NL80211_SB_DFS_BW_20,
+	NL80211_SB_DFS_BW_40,
+	NL80211_SB_DFS_BW_80,
+	NL80211_SB_DFS_BW_FULL,
+};
+
 #endif /* __LINUX_NL80211_H */
-- 
2.10.1

