# HG changeset patch
# Parent 3b7087a0594896134a332a1c28d78b5d33c7c0d6

diff --git a/arch/mips/lantiq/grx500_bootcore/Makefile b/arch/mips/lantiq/grx500_bootcore/Makefile
--- a/arch/mips/lantiq/grx500_bootcore/Makefile
+++ b/arch/mips/lantiq/grx500_bootcore/Makefile
@@ -1,5 +1,5 @@
-obj-y := prom.o clk.o irq.o sysctrl.o early_printk.o ngi.o timer.o
-
+obj-y := prom.o clk.o irq.o sysctrl.o ngi.o timer.o
+obj-$(CONFIG_SERIAL_GRX500_BOOTCORE_CONSOLE) += early_printk.o
 ifdef CONFIG_RESET_IAP
 obj-y += interaptiv_reset.o
 endif
diff --git a/arch/mips/lantiq/grx500_bootcore/prom.c b/arch/mips/lantiq/grx500_bootcore/prom.c
--- a/arch/mips/lantiq/grx500_bootcore/prom.c
+++ b/arch/mips/lantiq/grx500_bootcore/prom.c
@@ -23,6 +23,7 @@
 
 
 #define SOC_GRX500      "GRX500"
+#define SOC_GRX350      "GRX350"
 #define COMP_GRX500     "lantiq,grx500"
 
 #ifdef CONFIG_USE_EMULATOR
@@ -135,10 +136,18 @@ void __init ltq_soc_detect(struct ltq_so
         case SOC_ID_GRX582:
         case SOC_ID_GRX583:
         case SOC_ID_GRX580:
+        case SOC_ID_GRX584:
+        case SOC_ID_GRX585:
                 i->name = SOC_GRX500;
                 i->type = SOC_TYPE_GRX500;
                 i->compatible = COMP_GRX500;
                 break;
+		 case SOC_ID_GRX3506:
+		 case SOC_ID_GRX3508:
+		        i->name = SOC_GRX350;
+			    i->type = SOC_TYPE_GRX500;
+			    i->compatible = COMP_GRX500;
+			    break;
 #ifdef CONFIG_USE_EMULATOR
         case SOC_ID_PD:
                 i->name = SOC_GRX500;
