From 1667350de8e19fa48477c12965c5859f2ecc6af1 Mon Sep 17 00:00:00 2001
From: pukitis <martins.pukitis@intel.com>
Date: Fri, 19 Jan 2018 14:30:02 +0200
Subject: [PATCH] 020-intel_net_wireless_add_intel_specific_to_nl80211

---
 include/uapi/linux/nl80211.h | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/include/uapi/linux/nl80211.h b/include/uapi/linux/nl80211.h
index d3cbe48..b10f124 100644
--- a/include/uapi/linux/nl80211.h
+++ b/include/uapi/linux/nl80211.h
@@ -1936,6 +1936,10 @@ enum nl80211_commands {
  *	attribute.
  * @NL80211_ATTR_NAN_MATCH: used to report a match. This is a nested attribute.
  *	See &enum nl80211_nan_match_attributes.
+ * @NL80211_ATTR_VENDOR_WDS: used to enable the wds vendor specific capability
+ * @NL80211_ATTR_STA_RSSI: Received signal strength indication (u32)
+ * @NL80211_ATTR_RADAR_BIT_MAP: Bit map of channel where radar was detected
+ * @NL80211_ATTR_SB_DFS_BW: Sub Band DFS new operation bandwidth
  *
  * @NL80211_ATTR_BSSID: The BSSID of the AP. Note that %NL80211_ATTR_MAC is also
  *	used in various commands/events for specifying the BSSID.
@@ -2341,6 +2345,11 @@ enum nl80211_attrs {
 
 	NL80211_ATTR_BSSID,
 
+	NL80211_ATTR_VENDOR_WDS,
+	NL80211_ATTR_STA_RSSI,
+	NL80211_ATTR_RADAR_BIT_MAP,
+	NL80211_ATTR_SB_DFS_BW,
+
 	/* add attributes here, update the policy in nl80211.c */
 
 	__NL80211_ATTR_AFTER_LAST,
@@ -2698,6 +2707,7 @@ enum nl80211_sta_info {
 	NL80211_STA_INFO_TID_STATS,
 	NL80211_STA_INFO_RX_DURATION,
 	NL80211_STA_INFO_PAD,
+	NL80211_STA_INFO_MAX_RSSI,
 
 	/* keep last */
 	__NL80211_STA_INFO_AFTER_LAST,
@@ -5128,4 +5138,12 @@ enum nl80211_nan_match_attributes {
 	NL80211_NAN_MATCH_ATTR_MAX = NUM_NL80211_NAN_MATCH_ATTR - 1
 };
 
+enum nl80211_sb_dfs_bw {
+	NL80211_SB_DFS_BW_NORMAL,
+	NL80211_SB_DFS_BW_20,
+	NL80211_SB_DFS_BW_40,
+	NL80211_SB_DFS_BW_80,
+	NL80211_SB_DFS_BW_FULL,
+};
+
 #endif /* __LINUX_NL80211_H */
-- 
2.10.1

