From d0231252bd8e559702fa6bc2cdfc7ea70e82ef79 Mon Sep 17 00:00:00 2001
From: grishche <a.grishchenko@intel.com>
Date: Tue, 6 Jun 2017 15:05:50 +0300
Subject: [PATCH 15/20] mac80211_dont_intersect_reg_domain

---
 net/wireless/reg.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/net/wireless/reg.c b/net/wireless/reg.c
index 8eedb15..7044311 100644
--- a/net/wireless/reg.c
+++ b/net/wireless/reg.c
@@ -1370,8 +1370,9 @@ get_reg_request_treatment(struct wiphy *wiphy,
 		if (reg_request_cell_base(lr))
 			return REG_REQ_IGNORE;
 
+		/* Don't do intersection between mac80211 and cfg80211 regdom */
 		if (lr->initiator == NL80211_REGDOM_SET_BY_COUNTRY_IE)
-			return REG_REQ_INTERSECT;
+			return REG_REQ_OK;
 		/*
 		 * If the user knows better the user should set the regdom
 		 * to their country before the IE is picked up
@@ -1438,7 +1439,6 @@ __regulatory_hint(struct wiphy *wiphy,
 	struct regulatory_request *lr;
 
 	treatment = get_reg_request_treatment(wiphy, pending_request);
-
 	switch (treatment) {
 	case REG_REQ_INTERSECT:
 		if (pending_request->initiator ==
-- 
2.10.1

