From 5937507b7678960f340f8eba4bd5f3d70bd9be3b Mon Sep 17 00:00:00 2001
From: Artur Saygin <artur.saygin@intel.com>
Date: Wed, 5 Oct 2016 10:23:30 -0700
Subject: [PATCH 493/639] fixing Kconfig typos

---
 arch/x86/Kconfig | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/arch/x86/Kconfig b/arch/x86/Kconfig
index 6699d32..759e666 100644
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -461,7 +461,7 @@ config NET_SUBSYSTEM_SYSFS
 # WiFi Proxy network device driver configuration
 #
 
-menu "Intel Puma7 wifi-proxy Support"
+menu "Intel Puma7 WiFi proxy support"
 config WIFI_PROXY
         bool "WiFi proxy network device driver"
         default n
@@ -479,7 +479,7 @@ config WIFI_PROXY_DEBUG
           Enable WiFi proxy FW debugging
 
 config WIFI_PROXY_DEBUG_BUFF_SIZE
-        int "Debug buffer size, number of debuf messages"
+        int "Debug buffer size, number of debug messages"
         default 256
         depends on WIFI_PROXY_DEBUG
         ---help---
-- 
2.10.1

