# HG changeset patch
# Parent eb464e7d937412ae3772e4c39562bfc422d6c2eb

diff --git a/include/uapi/linux/atm.h b/include/uapi/linux/atm.h
--- a/include/uapi/linux/atm.h
+++ b/include/uapi/linux/atm.h
@@ -130,6 +130,11 @@
 #define ATM_ABR		4
 #define ATM_ANYCLASS	5		/* compatible with everything */
 
+#define ATM_VBR_NRT     ATM_VBR
+#define ATM_VBR_RT      6
+#define ATM_UBR_PLUS    7
+#define ATM_GFR         8
+
 #define ATM_MAX_PCR	-1		/* maximum available PCR */
 
 struct atm_trafprm {
